packages/aws-cdk/lib/api/hotswap/ecs-services.ts

```diff
@@ -195,11 +253,93 @@ async function prepareTaskDefinitionChange(
     // otherwise, familyNameOrArn is just the simple name evaluated from the CloudFormation template
     : familyNameOrArn;
   // then, let's evaluate the body of the remainder of the TaskDef (without the Family property)
-  return {
-    ...await evaluateCfnTemplate.evaluateCfnExpression({
+
+  let evaluated;
+  let hotswappableWithMerge = false;
+  try {
+    evaluated = await evaluateCfnTemplate.evaluateCfnExpression({
       ...(taskDefinitionResource ?? {}),
       Family: undefined,
-    }),
-    Family: family,
+    });
+  } catch (e) {
+    const result = await deepCompareContainerDefinitions(
+      evaluateCfnTemplate,
+      change.oldValue.Properties?.ContainerDefinitions ?? [],
+      change.newValue.Properties?.ContainerDefinitions ?? []);
+    if (result === false) {
+      throw e;
+    } else {
+      evaluated = {
+        // return a partial definition that should be merged with the current definition
+        ContainerDefinitions: result,
+      };
+      hotswappableWithMerge = true;
+    }
+  }
+
+  return {
+    taskDefinition: {
+      ...evaluated,
+      Family: family,
+    },
+    family,
+    hotswappableWithMerge,
   };
 }
+
+// return false if the new container definitions cannot be hotswapped
+// i.e. there are changes containing unsupported tokens or additions/removals of properties
+// If not, we return the properties that should be updated (they will be merged with the deployed task definition later)
+async function deepCompareContainerDefinitions(evaluateCfnTemplate: EvaluateCloudFormationTemplate, oldDefinition: any[], newDefinition: any[]) {
+  const result: any[] = [];
+  // schema: https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_ContainerDefinition.html
+  if (oldDefinition.length !== newDefinition.length) {
+    // one or more containers are added or removed
+    return false;
+  }
+  for (const i in oldDefinition) {
+    result.push({});
+    const prev = oldDefinition[i];
+    const next = newDefinition[i];
+
+    if (!deepCompareObject(Object.keys(prev), Object.keys(next))) {
+      // one or more fields are added or removed
+      return false;
+    }
+    for (const key of Object.keys(prev)) {
+      // compare two properties first
+      if (deepCompareObject(prev[key], next[key])) {
```

Comment:
Doesn't this need to be recursive? Or we're purposely not supporting the case where a simple env variable got added?  (That's fine, but I'd like that called out in comments)

Reply:
No, I'm afraid it overly complicates the code, while not adding much value to the actual hotswap experience. I added a comment: `We don't recurse properties to simplify the update logic. It should still cover most hotswap use cases.`

We might revisit this decision when someone actually needed the deep comparison :)

Reply:
On second thought, I changed it to recurse anyway to make the logic more reusable.